### DISCUSSION TITLE
Restrict custom assertion to a specific (reference) type while keeping existing object assertions

### DISCUSSION THREAD

**Discussion by andreivanea** [fluentassertions/fluentassertions#1999] [Q&A] [answered]
Hi there,

I've trying to write a custom assertion for a class. I've tried following [this discussion](https://github.com/fluentassertions/fluentassertions/discussions/1834) about extending existing assertions but I didn't get the results that I've hoped for. Consider the following code.

```csharp 
public class MyClass { }

public static class MyClassExtensions
{
    public static AndConstraint<ObjectAssertions> BeTheBestClassEver(this ObjectAssertions parent, 
         string because = "", params object[] becauseArgs)
    {
         // ...
    }
}
```

This kind of works. I don't lose existing object assertions on my class (such as `BeEquivalentTo`) and I'm able to use the assertion like this:

```csharp
new MyClass().should().BeTheBestClassEver();
```

The problem is that I can also do this:

```csharp
new { }.should().BeTheBestClassEver();
```

How can I restrict this custom assertion to my specific type, while keeping existing assertion?  I've looked into using the generic version `ObjectAssertions<TSubject, TAssertions>` and specifying my type as `TSubject` but I'm not sure what do with `TAssertions` as I get the following error: `The type 'FluentAssertions.Primitives.ObjectAssertions' cannot be used as type parameter 'TAssertions' in the generic type or method 'ObjectAssertions<TSubject, TAssertions>'. There is no implicit reference conversion from 'FluentAssertions.Primitives.ObjectAssertions' to 'FluentAssertions.Primitives.ObjectAssertions<MyClass, FluentAssertions.Primitives.ObjectAssertions>'`.

What's the correct way to implement a custom assertion that can be applied on a specific type?

**Comment by andreivanea** [comment, marked as answer]
This might have been the wrong approach here. I only "explored" alternative solutions because following [the official documentation on creating custom assertions](https://fluentassertions.com/extensibility/#building-your-own-extensions) did not yield the expected result. What I had written initially was:

```csharp
public class MyClassAssertions : ReferenceTypeAssertions<MyClass, MyClassAssertions>
{
    public MyClassAssertions(MyClass instance)  : base(instance) { }

    public AndConstraint<MyClassAssertions> BeTheBestClassEver(string because = "", params object[] becauseArgs)
    {
        // ...
    }
}
```

This meant that all assertions defined in `ObjectAssertions` were basically removed from my class. To get keep existing assertions *and* get my new, custom assertion *for my type only*, I've replaced `ReferenceTypeAssertions` with `ObjectAssertions`:

```csharp
public class MyClassAssertions : ObjectAssertions<MyClass, MyClassAssertions>
{
}
```

My conclusion is that extending `ObjectAssertions` instead of `ReferenceTypeAssertions` would be preferred in most cases.

**Reply by dennisdoomen** [reply]
This is also the correct approach. And you obviously need an extension method named `Should` on `MyClass` that returns an instance of `MyClassAssertions`
